### PR TITLE
Milestone3 - LOGIC OF THE GAME

### PR DESCRIPTION
**Pull request by Anvi98**
In This milestone, we applied the OOP concept to encapsulate our code. We coded the logic to find who can be the winner and when a TIE state can be raised.
Class created :
- Game class
- TictacToe Class

### CODE REVIEW

File: lib/logic.rb
Comment (e71az) [inline lib/logic.rb:53]:
```markdown
- [ ] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

```

Comment (eliasecasta) [inline lib/logic.rb]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: bin/main
Comment (eliasecasta) [inline bin/main]: - [x] All game-logic related resources (classes and/or modules) are kept in the `lib` directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

File: bin/main
Comment (eliasecasta) [inline bin/main]: - [x] The code should respect the single responsibility rule for methods (i.e. each method does exactly one thing). ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/single_responsibility.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

If I am not mistaken, you are taking a user input and also checking for a winner with the arbiter, please divide those steps into 2 different functions. (If I am mistaken, ignore this comment.)

File: bin/main
Comment (eliasecasta) [inline bin/main]: - [x] The student uses attr_readers/attr_writers. Only uses attr_accessors if they really need it and must be aware that by using it they allow writing that property from outside the class. That variable can be set to anything. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

Modify this to attr_reader and attr_writers unless you are sure that you need accessors please.

File: lib/logic.rb
Comment (eliasecasta) [inline lib/logic.rb]: - [x] The student uses attr_readers/attr_writers. Only uses attr_accessors if they really need it and must be aware that by using it they allow writing that property from outside the class. That variable can be set to anything. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

Modify this to attr_reader and attr_writers unless you are sure that you need accessors please.

File: lib/logic.rb
Comment (eliasecasta) [inline lib/logic.rb:73]: - [x] The student uses attr_readers/attr_writers. Only uses attr_accessors if they really need it and must be aware that by using it they allow writing that property from outside the class. That variable can be set to anything. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

Modify this to attr_reader and attr_writers unless you are sure that you need accessors please.

**Review by eliasecasta — CHANGES_REQUESTED** [review]
# Status:  Changes Required ♻️

Hello @Anvi98 and @mrigorir,

Remember to create more descriptive commit messages, for example, your "work on milestone3" and "fix issues on milestone3" are not descriptive enough. Take that into account for all your future commit messages as developers.

Good job with your Readme, just a small thing to point out: 

- [x] Your readme is pointing to ruby bin/main.rb but your file is missing the .rb extension, please correct that in your readme.

Great project, it was a really strong submission but there are some things that need fixing before it can be approved:

- [x] Remove the puts that you used from the logic file, they are only allowed in the main one.

## Behaviour:

- [x] The application saves players move and displays status of the board after detecting an invalid player input. (currently, if you select one spot twice and then select another one, the player sign won't be updated in the board)

- [x] Currently if I make a wrong move, your game starts to bug out and declares wrong winners:  (I never selected spot 7)
image

## Implementation details:

- [x] All game-logic related resources (classes and/or modules) are kept in the `lib` directory.
- [x] The code should respect the single responsibility rule for methods.
- [x] The student uses attr_readers/attr_writers.

[OPTIONAL]:

- [x] Ruby methods automatically return the last evaluated statement therefore, there is no need for explicit declaration `return`.
- [x] No hard coded values like your @indexes_win variable in your logic file.
- [x] The Game class has two attributes for the players: @player1 and @player2. Ideally, there should be a Player class and @players attribute in Game class which is an array that stores the Player instances.

Please correct those issues and send another review request, good luck, and well done!

File: lib/game.rb
Comment (Berabjesus) [inline lib/game.rb:24]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright)
](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)

File: lib/game.rb
Comment (Berabjesus) [inline lib/game.rb:41]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright)
](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)

File: lib/game.rb
Comment (Berabjesus) [inline lib/game.rb:54]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright)
](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)

File: lib/game.rb
Comment (Berabjesus) [inline lib/game.rb:93]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright)
](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)

File: lib/game.rb
Comment (Berabjesus) [inline lib/game.rb:104]: - [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright)
](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)

File: lib/logic.rb
Comment (Berabjesus) [inline lib/logic.rb:3]: - [x] Remove unnecessary attr_read/accessor values that you won't use anywhere else.

**Review by Berabjesus — CHANGES_REQUESTED** [review]
## Status: REQUIRED CHANGES :warning: :construction: 
Hello @Anvi98 and @mrigorir such a remarkable project you made 👏. You have also corrected some of the suggestions given to you by your previous reviewer 👍. 
### However, you have a few changes to carry out before we proceed so please take a look at the things you have to work on and their respective review comments below.

- [x] No `puts/print` and `gets.chomp` outside `main.rb` file. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png) from [Pathwright)
](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)

- [x] The application does not exit/stop due to invalid input - it informs the player about it and asks for their move again. (At the moment, the program crashes if i hit enter multiple times with the error below).
image

- [x] The student uses attr_readers/attr_writers. Only uses attr_accessors if they really need it. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/attr_accessor.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

**Please have a look at the specific comments below**

 **Please, do not open a new Pull Request for subsequent reviews. You should use the same Pull Request submitted for the first review unless it is requested otherwise.**

#### **If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

**Review by mke2111 — COMMENTED** [review]
# STATUS: CHANGES REQUIRED 🔴 ♻️ 

## Hey Happy Coders, 👋🏽 
**This is a great project you have going on, well-done lads and thanks for implementing the changes requested earlier on**. However, there are a few issues to fix to get approved.

### REQUIRED CHANGES: ⚠️ 

- [x] All game-logic related resources (classes and/or modules) are kept in the lib directory. [Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/)
> Find means of keeping the statements `puts` and `gets` only in the `main` file without adding the `class Game` in the `main` file.

- [x] In the readme file, the command to run the main file should be `bin/main`. Or you could as well `cd TicTacToe/bin` then run `main`.
- [x] [OPTIONAL] You may choose to create a separate file for the Player Class, still within the lib directory.

### Kindly make all the requested changes and submit another CRR. 💯 

**Cheers 🍹 and Happy coding 💻** 

## Ping me on slack for a chit-chat 🆘  Roy Mukuye Or mention me in this PR

**Comment by mimipeshy** [comment]
### STATUS: CHANGES REQUESTED 🔴 ♻️ 

Hello, team @Anvi98 and @mrigorir 👋🏾 👋🏾 this is a great effort you have put into this milestone. You have implemented the TSE changes required of you. Your project runs very well, that is a good job. However, before approval, there is a TSE comment you didn't work on as I have noted before

- [x] There is still some puts statement inside `lib/game.rb` file . If you have to use puts statements kindy use them only in `main.rb` file
([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone2/main_rb.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

 **Kindly do not merge a PR unless it's been approved by a TSE or it's a requirement.**
That's all :smile: kindly submit another review when done.

_Feel free to reach out to me on slack if you have any questions or something is unclear, or just comment on this PR and tag my name  (@mimipeshy) and I will be glad to help_
Regards
Peris Ndanu

**Review by mricanho — COMMENTED** [review]
# Status: ✔️ Project Approved ✔️ 

Hello, @Anvi98 @mrigorir, good job on this milestone! 🔥 

You worked hard in all the instructions that the previous Code Reviewers give you👍🏽 

ok ok

[ OPTIONAL ]
- Try refactoring your code, if you cut your methods into small pieces it is easier to reuse and avoid headaches.

Now you can merge this branch 🚀 

You can contact me on slack if you have any questions, just search for me as @ Miguel Ricaño.

### 💻 👩🏽‍💻  Happy Coding!  👨🏽‍💻 🖥️

**Pull request merged**
